### ISSUE TITLE
omc doesn't report errors from .mos input

### ISSUE THREAD

**Issue by jdpipe** [OpenModelica/OpenModelica#10043]
### Description

I have a script `myscript.mos` containing text as follows:
```
justNonsense()
```

When I run this script through `omc`,
```
omc myscript.mos
```
I get a single newline as output, with no error message. Also, the exit code is zero. There is no explicit way to check for an error in the .mos input!!

### Expected Behavior

I would expect to be able to check for both an error message and an exit code in a case like this. Normally omc gives such verbose output! But this time nothing.

![image](https://user-images.githubusercontent.com/87423/212213027-6446da70-176d-46df-b3b0-954f82d82769.png)

So obviously I can test for this 'true' output in the second case, but really an error should be reported for the first case, and that's not happening, and should be fixed, I feel.

### Version and OS

omc 1.20.0 (release) on Ubuntu 20.04.

See also #10040.

**Comment by casella** [comment]
@jdpipe did you try `justNonsense();getErrorString()` ?

**Comment by jdpipe** [comment]
I didn't try that. The point is that I won't know ahead of time that I have made a mistake, so why would I request an error string to be written out...?

**Comment by perost** [comment]
The issue is that a scripting calls never give errors, they just return an expression. So the normal behaviour when an error occurs is that the scripting call returns some expression indicating an error (usually `fail()` or `false`), and then `getErrorString()` can be used to get the error buffer as a string.

But in this case we don't get anything at all when calling a non-existent function, so that should be fixed at least. The question is what should be done in that case. I guess the most idiomatic solution would be to return `fail()` in that case, unless someone has a better idea? The only issue with that solution is that it doesn't differentiate between a non-existent function and a function that exists but fails. But maybe that's not an important distinction.

**Comment by bilderbuchi** [comment]
> And regarding the exit code, the problem is that a script doesn't really fail. Individual statements can fail, but the script execution continues regardless.

Is that intended, or by accident? Do you consider this generally valid, or a Modelica scripting idiosyncracy?

If I run a nonsense python script (where an individual statement fails), e.g.
```
python -c "invalid_command(); print('foo')"
```
I get an error message
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'invalid_command' is not defined
```
and the exit code is `1`.
OTOH, for bash scripts you need to `set -e` to abort when encountering an "erroring" statement, so there's that.

**Comment by bilderbuchi** [comment]
Personally, I've always found this need to tack `getErrorString()` onto commands to actually notice when they error _quite_ annoying (and this is partially why I avoid scripting Modelica using mos). 
Is there no other possibility, e.g. to use `stderr` to print errors by default, and `stdout` for the normal/intended output?

**Comment by casella** [comment]
@perost I guess if a script calls a non-existent function should return an error. However, given the way function calls are handled i the API, maybe returning false() and then putting some error string like "Function xxx does not exist" to be returned by getErrorString() would be OK.

**Comment by jdpipe** [comment]
I agree that all errors should appear in the `stderr` stream immediately they occur. I feel that this is the only logical behaviour. What is the point in swallowing errors in a low-level tool like `omc`, and expecting the user to know when they need to know? One can always suppress them if desired `omc myscript.mos 2>/dev/null`.

I think that after each command in a `.mos` script, if there has been error output, then it should be possible to trigger the script to stop. This could be with an extra flag, like `-e`, as suggested.

But I think that there could possibly be an architectural issue here, which is that it seems that there is no clear distinction between functions returning a value (like Python, which needs an exception mechanism in parallel) and functions return an error code (like the 'true'/'false' of `loadModel`, or the error code of `system`). So it is hard to code in such a way that unanticipated problems are brought naturally to the attention of the coder. The `getErrorString()` thing seems like a klugde to cover up for this lack of a better structure of error handling. Sometimes you don't mind about an error, because you have an alternative action that you want to run. So if we want to exit after any error, we also need a way in the language of catching errors and not lettting them bubble up.

**Comment by perost** [comment]
In #10059 I've now changed it so that an expression that couldn't be evaluated will return `fail()` instead of just an empty string. I'm not sure if it's really much of an improvement, but I'm curious to see how it impacts our testsuite.

But regarding other changes like error codes and better handling of error messages, the issue is that the current scripting implementation is around 20 years old (even predating MetaModelica) and not so much designed as organically grown. Making any changes to how errors are handled is unfortunately very hard since the code is rather convoluted, not to mention that we have a lot of scripts that we really don't want to break.

Adding an extra flag to make scripts stop on errors might be possible, but it might be hard to actually make it work reliably due to how the scripting is implemented. I'm also not sure which errors should be considered hard errors, there's so much more that can go wrong than just calling a non-existent function.

So I think solving these issues properly might require breaking backward compatibility, or possibly even implementing a new scripting language that's better designed. But that's not going to happen any time soon I'm afraid, so for now we're unfortunately stuck with e.g. `getErrorString()`.

**Comment by perost** [comment]
As expected #10059 broke a lot of our test cases, but it looks like it's mostly that their expected output needs to be updated. I'm a bit worried that it might break scripts in unexpected ways though, since we use scripts for e.g. our build processes.

Maybe @adrpo or @sjoelund has some opinions on this?

**Comment by sjoelund** [comment]
Does OMPython handle this?

**Comment by casella** [comment]
@jdipipe I agree with you comment somewhere else that if you want to do serious scripting, you better use Python or Matlab or some other well-developed language. We have OMPyton and OMMatlab interfaces, and we are constantly trying to improve them. OpenModelica scripting is really meant for low-level service tasks, not for serious software development.
